package/kedro_viz/server.py

```diff
@@ -317,90 +324,143 @@ def _construct_layer_mapping(catalog):
+            pipeline_key,
+            pipeline,
+            catalog,
+            nodes,
+            node_dependencies,
+            tags,
+            edges_list,
+            nodes_list,
+        )
+
+    # sort tags
+    sorted_tags = [{"id": tag, "name": _pretty_name(tag)} for tag in sorted(tags)]
+    # sort layers
+    sorted_layers = _sort_layers(nodes, node_dependencies)
+    return {
+        "nodes": nodes_list,
+        "edges": edges_list,
+        "tags": sorted_tags,
+        "layers": sorted_layers,
+        "pipelines": pipelines_list,
+    }
+
+
+# pylint: disable=too-many-locals,too-many-arguments
+def format_pipeline_data(
+    pipeline_key: str,
+    pipeline: "Pipeline",  # noqa: F821
+    catalog: "DataCatalog",  # noqa: F821
+    nodes: Dict[str, dict],
+    node_dependencies: Dict[str, dict],
+    tags: Set[str],
+    edges_list: List[dict],
+    nodes_list: List[dict],
+) -> None:
+    """Format pipeline and catalog data from Kedro for kedro-viz.
+
+    Args:
+        pipeline_key: key value of a pipeline object (e.g "__default__").
+        pipeline: Kedro pipeline object.
+        catalog:  Kedro catalog object.
+        nodes: Dictionary of id and node dict.
+        node_dependencies: Dictionary of id and node dependencies.
+        edges_list: List of all edges.
+        nodes_list: List of all nodes.
+
+    """
     # keep_track of {data_set_namespace -> set(tags)}
     namespace_tags = defaultdict(set)
     # keep track of {data_set_namespace -> layer it belongs to}
     namespace_to_layer = {}
-    all_tags = set()
 
     dataset_to_layer = _construct_layer_mapping(catalog)
 
     for node in sorted(pipeline.nodes, key=lambda n: n.name):
         task_id = _hash(str(node))
-        all_tags.update(node.tags)
-        nodes[task_id] = {
-            "type": "task",
-            "id": task_id,
-            "name": getattr(node, "short_name", node.name),
-            "full_name": getattr(node, "_func_name", str(node)),
-            "tags": sorted(node.tags),
-        }
-        sorted_nodes.append(nodes[task_id])
+        tags.update(node.tags)
+        if task_id not in nodes:
+            nodes[task_id] = {
+                "type": "task",
+                "id": task_id,
+                "name": getattr(node, "short_name", node.name),
+                "full_name": getattr(node, "_func_name", str(node)),
+                "tags": sorted(node.tags),
+                "pipelines": [pipeline_key],
+            }
+            nodes_list.append(nodes[task_id])
+        else:
+            nodes[task_id]["pipelines"].append(pipeline_key)
 
         for data_set in node.inputs:
             namespace = data_set.split("@")[0]
             namespace_to_layer[namespace] = dataset_to_layer.get(data_set)
             namespace_id = _hash(namespace)
-            edges.append({"source": namespace_id, "target": task_id})
+            edge = {"source": namespace_id, "target": task_id}
+            if edge not in edges_list:
```

Comment:
I saw that @limdauto had some concerns about performance here. Can you use `set` for `edges_list`?

Reply:
I think managing dict in a set seems to be a bit more verbose (that's why `edges_list` was a list [originally](https://github.com/quantumblacklabs/kedro-viz/blob/develop/package/kedro_viz/server.py#L342) )? https://stackoverflow.com/questions/34097959/add-a-dictionary-to-a-set-with-union

Reply:
Well for `set`, it will be the same just in the opposite direction. `in` is O(1) but `add` is O(n).  But maybe amortized cost will be cheaper since we are going to `add` less frequently than checking `in`.